### PR TITLE
Set the maximum width of categorical bars & boxes to category unit - case of categoryarray

### LINKED ISSUE #5731: bar widths and categoryarray
In this pen https://codepen.io/nicolaskruchten/pen/eYvbyzJ I believe the bars should not overlap the "b" category.

Per https://github.com/plotly/plotly.js/pull/5072#issuecomment-673706563 I think that on categorical axes, the bar "span" (i.e. the width before taking into account bargap/bargroupgap) should be 1 always, not just when using overlay.

Note: this should apply to box and violin as well.

### PR DESCRIPTION
**Pull request by archmoj**
Fixes #5731.

@plotly/plotly_js

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
Behaviour looks good.

**Comment by nicolaskruchten** [comment]
Interestingly, box and violin don't seem to suffer from this problem even on master ... normal?

**Comment by archmoj** [comment]
> Interestingly, box and violin don't seem to suffer from this problem even on master ... normal?

Positive. Thanks for checking.

File: src/traces/bar/sieve.js [hunk 1]
```diff
@@ -40,14 +40,17 @@ function Sieve(traces, opts) {
...
-    });
+    var dv = distinctVals(positions);
 
     this.distinctPositions = dv.vals;
     if(dv.vals.length === 1 && width1 !== Infinity) this.minDiff = width1;
     else this.minDiff = Math.min(dv.minDiff, width1);
 
+    var type = (opts.posAxis || {}).type;
+    if(type === 'category' || type === 'multicategory') {
+        this.minDiff = Math.min(1, this.minDiff);
```
Comment (alexcjohnson) [inline src/traces/bar/sieve.js:51]: This looks like a great solution, and simpler than the previous logic. My only question is whether there's actually a benefit of doing `Math.min` vs just `minDiff = 1` in the (multi)category case? There's no way to specify input data that will result in non-integer positions here, is there?

Comment (archmoj) [reply src/traces/bar/sieve.js]: Good call. Set to one in https://github.com/plotly/plotly.js/pull/5732/commits/6e6bfa693492867bd9a4c35d2ff75a6755fede8c.

File: src/traces/box/cross_trace_calc.js [hunk 2]
```diff
@@ -60,9 +60,10 @@ function setPositionOffset(traceType, gd, boxList, posAxis) {
     if(!pointList.length) return;
 
     // box plots - update dPos based on multiple traces
-    var boxdv = Lib.distinctVals(pointList, {
-        unitMinDiff: posAxis.type === 'category' || posAxis.type === 'multicategory'
-    });
+    var boxdv = Lib.distinctVals(pointList);
+    if(posAxis.type === 'category' || posAxis.type === 'multicategory') {
+        boxdv.minDiff = Math.min(1, boxdv.minDiff);
```
Comment (alexcjohnson) [inline src/traces/box/cross_trace_calc.js:65]:
```suggestion
        boxdv.minDiff = 1;
```
Same tweak here :)

**Review by alexcjohnson — APPROVED** [review]
💃  after considering https://github.com/plotly/plotly.js/pull/5732/files#r653694361

**Pull request merged**